src/core/api.rs

```diff
@@ -174,168 +460,157 @@ impl Covercrypt {
-        public_key: &MasterPublicKey,
-        encryption_policy: &AccessPolicy,
-        metadata: Option<&[u8]>,
-        authentication_data: Option<&[u8]>,
-    ) -> Result<(SymmetricKey<SEED_LENGTH>, Self), Error> {
-        let (symmetric_key, encapsulation) =
-            cover_crypt.encaps(policy, public_key, encryption_policy.clone())?;
-
-        let encrypted_metadata = metadata
-            .map(|bytes| cover_crypt.dem_encrypt(&symmetric_key, bytes, authentication_data))
-            .transpose()?;
-
-        Ok((
-            symmetric_key,
-            Self {
-                encapsulation,
-                encrypted_metadata,
-            },
-        ))
+        ap: &str,
+        ad: Option<&[u8]>,
+        plaintext: &[u8],
+    ) -> Result<(Encapsulation, Vec<u8>), Error> {
+        let (seed, enc) = CovercryptKEM::encaps(self, mpk, policy, ap)?;
+        let mut sym_key = Secret::<KEY_LENGTH>::default();
+        kdf256!(&mut sym_key, &seed);
+        let res = E::encrypt(&sym_key, plaintext, ad, &self.rng)?;
+        Ok((enc, res))
     }
 
-    /// Decrypts the header with the given user secret key.
+    /// Attempts decrypting the given ciphertext using the Covercrypt KEM and the DEM.
     ///
-    /// The nonce used is extracted from the encapsulation tag.
+    /// Attempts opening the Covercrypt encapsulation. If it succeeds, decrypts
+    /// the ciphertext using the DEM with the encapsulated key and the given
+    /// authentication data. Returns ‘None‘ otherwise.
     ///
-    /// - `cover_crypt`         : `Covercrypt` object
-    /// - `usk`                 : `Covercrypt` user secret key
-    /// - `authentication_data` : authentication data used in the DEM encryption
-    pub fn decrypt(
+    /// # Error
+    ///
+    /// Returns an error if the access policy is not valid.
+    fn decrypt(
         &self,
-        cover_crypt: &Covercrypt,
         usk: &UserSecretKey,
-        authentication_data: Option<&[u8]>,
-    ) -> Result<Option<CleartextHeader>, Error> {
-        if let Some(symmetric_key) = cover_crypt.decaps(usk, &self.encapsulation)? {
-            let metadata = self
-                .encrypted_metadata
-                .as_ref()
-                .map(|ciphertext| {
-                    cover_crypt.dem_decrypt(&symmetric_key, ciphertext, authentication_data)
-                })
-                .transpose()?;
-            Ok(Some(CleartextHeader {
-                symmetric_key,
-                metadata,
-            }))
-        } else {
-            Ok(None)
-        }
+        ad: Option<&[u8]>,
+        ciphertext: &[u8],
+        enc: &Encapsulation,
+    ) -> Result<Option<Vec<u8>>, Error> {
+        let seed = CovercryptKEM::decaps(self, usk, enc.clone());
+        seed.map(|seed| {
+            let mut sym_key = Secret::<KEY_LENGTH>::default();
+            kdf256!(&mut sym_key, &seed);
+            let plaintext = E::decrypt(&sym_key, ad, ciphertext)?;
+            Ok(plaintext)
```

Comment:
Maybe the error conversion can be inferred:
```suggestion
            E::decrypt(&sym_key, ad, ciphertext).into()
```